### ISSUE TITLE
[Corporate Sponsor needed] Refactor to modern standards

### ISSUE THREAD

**Issue by splitbrain** [splitbrain/dokuwiki-plugin-bureaucracy#303]
This plugin was originally written by me to handle simple mail forms only. During a research project in 2011 the plugin was enhanced to add the template mode. This was part of my day job at CosmoCode. Since then the plugin has been extended and modified a lot to fit individual customer requirements and special use cases. It has become one of the most important plugins in corporate contexts.

Unfortunately its age and history shows in the code. The plugin has many problems:

  * low test coverage
  * outdated coding style
  * overly complex methods
  * still uses the deprecated form mechanism (which will go away in the future)
  * grown architecture that makes it hard to extend the plugin
  * lacking documentation

**In other words, this plugin needs some serious refactoring to make it future proof.** 

I estimate that a refactoring of the plugin (not adding any significant new features) will take about a week (40h) of work. I personally can not invest that much time into it.

So if your business relies on this plugin, please consider [hiring CosmoCode](https://www.cosmocode.de/en/leistungen/wiki/dokuwiki/) to do this.

**Comment by iainhallam** [comment]
Could we crowdfund the refactor? How much do you think it would cost, @splitbrain?

**Comment by splitbrain** [comment]
Cost would be about 20k EUR. For hiring CosmoCode a single client is needed. Of course that client could crowdfund...

**Comment by dregad** [comment]
> Edit: fixed the amount - I'm bad at math ;-)

It was a nice daily rate... I was going to ask if you were hiring 😉

Still way out of our budget though...

**Comment by listerr** [comment]
I could potentially put up half of the funds, maybe we could crowdfund the rest?

bureaucracy is super useful, but showing its age and is a bit inflexible in places. 

(And it could be better at styling.)

I made a couple of small hacks to make it do what I need it to do. (Redirect to a different page after the form is submitted.)

I'll have to look a bit more into the dokuwiki form mechanism. Perhaps worth looking at something like [Former](https://github.com/formers/former/wiki/Usage-and-Examples) rather then having to reinvent an entire template engine / form builder.

Let me know if interested!

R.

**Comment by splitbrain** [comment]
For everyone interested in crowd funding this: an alternative would be to have your company join the [DokuWiki Business Plugin Partner Program](https://www.cosmocode.de/en/services/wiki/dokuwiki-partnership/).
